salt: Generate custom kubeconfig for Salt Master

gdemonet opened the pull request:
Instead of using the `/etc/kubernetes/admin.conf` file which points to a
specific master's `kube-apiserver` instance, we generate another one
dedicated to Salt Master, and configure it to point to the local
`apiserver-proxy` (which can then route to other masters if the local
one isn't available).

In addition, this kubeconfig generates its own certificate, which could
later map to another group (and thus, other (Cluster)Role(s) than the
current "system:masters").

Note that we reduce the unneeded `/etc/kubernetes` mount to
`/etc/kubernetes/pki` (for SA signing key and etcd encryption key) in
both Salt Master and Salt API containers (SaltAPI didn't need it since
41ba749).

Fixes: #2533

gdemonet committed:
salt: Generate custom kubeconfig for Salt Master

Instead of using the `/etc/kubernetes/admin.conf` file which points to a
specific master's `kube-apiserver` instance, we generate another one
dedicated to Salt Master, and configure it to point to the local
`apiserver-proxy` (which can then route to other masters if the local
one isn't available).

In addition, this kubeconfig generates its own certificate, which could
later map to another group (and thus, other (Cluster)Role(s) than the
current "system:masters").

Note that we reduce the unneeded `/etc/kubernetes` mount to
`/etc/kubernetes/pki` (for SA signing key and etcd encryption key) in
both Salt Master and Salt API containers (SaltAPI didn't need it since
41ba749).

Fixes: #2533

bert-e commented:
# Hello gdemonet,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/2533-salt-master-kubeconfig` into
`development/2.4`
* [w/2.5/bugfix/2533-salt-master-kubeconfig](https://github.com/scality/metalk8s/compare/development/2.5...w/2.5/bugfix/2533-salt-master-kubeconfig) will be merged into `development/2.5` 
* [w/2.6/bugfix/2533-salt-master-kubeconfig](https://github.com/scality/metalk8s/compare/development/2.6...w/2.6/bugfix/2533-salt-master-kubeconfig) will be merged into `development/2.6` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @nootal

* @alexandre-allard-scality

* @lucieleonard

* @sayf-eddine-scality

* @slaperche-scality

* @nicolast

* @gdemonet

* @carlito-scality

* @thomasdanan

* @jbertran

* @teddyandrieux

* @cmonfort

* @monpote

bert-e commented:
# History mismatch



Merge commit #a51669ec69d0c4a1277168c0fecaa5407c065efd on the integration branch
`w/2.5/bugfix/2533-salt-master-kubeconfig` is merging a branch which is neither the current
branch `bugfix/2533-salt-master-kubeconfig` nor the development branch
`development/2.5`.

It is likely due to a rebase of the branch `bugfix/2533-salt-master-kubeconfig` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

gdemonet committed:
salt: Fix message if missing SA private key

gdemonet commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/2533-salt-master-kubeconfig` into
`development/2.4`
* [w/2.5/bugfix/2533-salt-master-kubeconfig](https://github.com/scality/metalk8s/compare/development/2.5...w/2.5/bugfix/2533-salt-master-kubeconfig) will be merged into `development/2.5` 
* [w/2.6/bugfix/2533-salt-master-kubeconfig](https://github.com/scality/metalk8s/compare/development/2.6...w/2.6/bugfix/2533-salt-master-kubeconfig) will be merged into `development/2.6` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @nootal

* @alexandre-allard-scality

* @lucieleonard

* @sayf-eddine-scality

* @slaperche-scality

* @nicolast

* @gdemonet

* @carlito-scality

* @thomasdanan

* @jbertran

* @teddyandrieux

* @cmonfort

* @monpote

bert-e commented:
# History mismatch



Merge commit #ee56b990cdf217e21cacfe853c6f4d88bacc1378 on the integration branch
`w/2.5/bugfix/2533-salt-master-kubeconfig` is merging a branch which is neither the current
branch `bugfix/2533-salt-master-kubeconfig` nor the development branch
`development/2.5`.

It is likely due to a rebase of the branch `bugfix/2533-salt-master-kubeconfig` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

gdemonet commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/2533-salt-master-kubeconfig` into
`development/2.4`
* [w/2.5/bugfix/2533-salt-master-kubeconfig](https://github.com/scality/metalk8s/compare/development/2.5...w/2.5/bugfix/2533-salt-master-kubeconfig) will be merged into `development/2.5` 
* [w/2.6/bugfix/2533-salt-master-kubeconfig](https://github.com/scality/metalk8s/compare/development/2.6...w/2.6/bugfix/2533-salt-master-kubeconfig) will be merged into `development/2.6` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @nootal

* @alexandre-allard-scality

* @lucieleonard

* @sayf-eddine-scality

* @slaperche-scality

* @nicolast

* @gdemonet

* @carlito-scality

* @thomasdanan

* @jbertran

* @teddyandrieux

* @cmonfort

* @monpote

gdemonet commented:
/wait

gdemonet commented:
This doesn't work with upgrade: obviously, haven't run the `metalk8s.salt.master.kubeconfig` formula yet...

Trying a refactor, moving this PR back to draft state.

gdemonet committed:
salt: Extract early-stage logic from `salt.master` states

We have too much logic in our script, lets move this logic in Salt
directly, add a state to deploy early-stage of the bootstrap.
Make the `salt.master` init state complet, meaning only few state of
`salt.master` must be run in case you are in local mode

gdemonet committed:
salt: Add good raise message in `metalk8s.minions_by_role`

In `metalk8s.minions_by_role` custom Salt module function if we are not
able to retrieve the list of Nodes from the pillar we just get a
KeyError without any information, this commit add a good raising message
in this specific case

TeddyAndrieux committed:
script: Use early stage bootstrap state in restore script

A state was added to handle bootstrap early stage deployment, use this
salt state in restore script

TeddyAndrieux committed:
salt: Do not upgrade kubelet when running apiserver-proxy state

We do no include kubelet state in apiserver-proxy state as it may result
to kubelet, containerd upgrade in case you call apiserver-proxy salt
state in upgrade process. This include is not needed here

TeddyAndrieux committed:
salt: Use `metalk8s_kubernetes` module in our custom roster

Before this commit we use directly Kubernetes python library in our
custom Salt roster, instead of it use directly our `metalk8s_kubernetes`
Salt module to interact with kubernetes

TeddyAndrieux committed:
salt: Add support for `Volume` in `metalk8s_kubernetes` Salt module

Add support for our custom CRD for `Volume` in our `metalk8s_kubernetes`
custom Salt module to manage kubernetes objects

TeddyAndrieux committed:
salt: Use `metalk8s_kubernetes` module in `metalk8s_nodes` ext_pillar

Before this commit we use directly Kubernetes python library in our
`metalk8s_nodes` external pillar, instead of it use directly our
`metalk8s_kubernetes` Salt module to interact with kubernetes

TeddyAndrieux committed:
script: Use `SALTENV` var everywhere in upgrade script

To be consistent use `SALTENV` variable as saltenv instead of compute
`metalk8s-$DESTINATION_VERSION` for some and use `SALTENV` directly for
others. In usual case `SALTENV` will be equal to
`metalk8s-$DESTINATION_VERSION` and it's checked by the upgrade
pre-check

TeddyAndrieux committed:
script: Call `bootstrap.local` in Upgrade, downgrade scripts

Since `metalk8s.roles.bootstrap` is no longer "local compatible" we need
to call only `metalk8s.roles.bootstrap.local` in Upgrade and Downgrade.
NOTE: In Downgrade we use the destination version state so if we are in
2.50 or <= 2.4.3 then we need to use the old `metalk8s.roles.bootstrap`
so an "if" was added in the Downgrade script to call the right state

TeddyAndrieux committed:
salt,script: Add `pre-` Upgrade and Downgrade for bootstrap

Add an orchestrate to handle what need to be done before upgrading or
downgrading the bootstrap node, to move the logic from bash script to
Salt

TeddyAndrieux committed:
salt,script: Add Salt master kubeconfig generation in upgrade,downgrade

Before downgrading or upgrade the bootstrap we need to create the new
kubeconfig for Salt master and since this kubeconfig use the
apiserver-proxy we need to (re-)configure apiserver-proxy on the
bootstrap node.
Also move the apiserver re-generation in pre-upgrade of the bootstrap as
we need apiserver to accept 127.0.0.1 to use the apiserver-proxy

TeddyAndrieux commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.






*The following options are set:* **wait**

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.5/bugfix/2533-salt-master-kubeconfig` with contents from `bugfix/2533-salt-master-kubeconfig`
and `development/2.5`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.5/bugfix/2533-salt-master-kubeconfig origin/development/2.5
 $ git merge origin/bugfix/2533-salt-master-kubeconfig
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.5/bugfix/2533-salt-master-kubeconfig
```

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/2533-salt-master-kubeconfig` into
`development/2.4`
* [w/2.5/bugfix/2533-salt-master-kubeconfig](https://github.com/scality/metalk8s/compare/development/2.5...w/2.5/bugfix/2533-salt-master-kubeconfig) will be merged into `development/2.5` 
* [w/2.6/bugfix/2533-salt-master-kubeconfig](https://github.com/scality/metalk8s/compare/development/2.6...w/2.6/bugfix/2533-salt-master-kubeconfig) will be merged into `development/2.6` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @tcarmet

* @thomasdanan

* @sayf-eddine-scality

* @lucieleonard

* @monpote

* @carlito-scality

* @cmonfort

* @nicolast

* @slaperche-scality

* @teddyandrieux

* @nootal

* @gdemonet

* @alexandre-allard-scality

TeddyAndrieux committed:
changelog: Add entry for salt-master dedicated kubeconfig

Add a changelog entry for the new kubeconfig generated for the Salt
master interaction with kubernetes apiserver

Sees: #2533

TeddyAndrieux commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.5/bugfix/2533-salt-master-kubeconfig` with contents from `bugfix/2533-salt-master-kubeconfig`
and `development/2.5`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.5/bugfix/2533-salt-master-kubeconfig origin/development/2.5
 $ git merge origin/bugfix/2533-salt-master-kubeconfig
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.5/bugfix/2533-salt-master-kubeconfig
```

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/2533-salt-master-kubeconfig` into
`development/2.4`
* [w/2.5/bugfix/2533-salt-master-kubeconfig](https://github.com/scality/metalk8s/compare/development/2.5...w/2.5/bugfix/2533-salt-master-kubeconfig) will be merged into `development/2.5` 
* [w/2.6/bugfix/2533-salt-master-kubeconfig](https://github.com/scality/metalk8s/compare/development/2.6...w/2.6/bugfix/2533-salt-master-kubeconfig) will be merged into `development/2.6` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @tcarmet

* @thomasdanan

* @sayf-eddine-scality

* @lucieleonard

* @monpote

* @carlito-scality

* @cmonfort

* @nicolast

* @slaperche-scality

* @teddyandrieux

* @nootal

* @gdemonet

* @alexandre-allard-scality

TeddyAndrieux committed:
salt: Add Salt master kubeconfig generation in restore

New Salt master kubeconfig is needed to interact with kubernetes
apiserver so we need to re-generate it when configuring the new
Salt master during restore script

gdemonet commented on salt/_pillar/metalk8s_nodes.py:
```diff
@@ -36,39 +26,43 @@ def node_info(node, ca_minion):
+        result['version'] = node['metadata']['labels'][VERSION_LABEL]
 
-    for (label, value) in node.metadata.labels.items():
+    for (label, _) in node['metadata']['labels'].items():
```
`.keys()`?

gdemonet commented on salt/_pillar/metalk8s_nodes.py:
```diff
@@ -36,39 +26,43 @@ def node_info(node, ca_minion):
+            apiVersion="v1",
+            kubeconfig=kubeconfig
+        )
+    except Exception as exc:  # pylint: disable=broad-except
```
It's an execution module, so I guess it should only raise a `CommandExecutionError`?

gdemonet commented on salt/_pillar/metalk8s_nodes.py:
```diff
@@ -156,21 +144,28 @@ def ext_pillar(minion_id, pillar, kubeconfig):
+                "Failed to retrieve NodeList: {!s}".format(exc)
+            ])
+        else:
+            log.info("Successfully retrieved nodes for ext_pillar")
```
I put this for debug purposes, maybe it should be removed or at least toned down to `debug` level

gdemonet reviewed (commented):
Few things to adjust, will take care of them tomorrow. Thanks a lot @TeddyAndrieux for taking care of this!

bert-e commented:
# History mismatch



Merge commit #5ee08d15d0dfa23cd18d01bf007ff4268dc36922 on the integration branch
`w/2.5/bugfix/2533-salt-master-kubeconfig` is merging a branch which is neither the current
branch `bugfix/2533-salt-master-kubeconfig` nor the development branch
`development/2.5`.

It is likely due to a rebase of the branch `bugfix/2533-salt-master-kubeconfig` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

gdemonet commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.5/bugfix/2533-salt-master-kubeconfig` with contents from `bugfix/2533-salt-master-kubeconfig`
and `development/2.5`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.5/bugfix/2533-salt-master-kubeconfig origin/development/2.5
 $ git merge origin/bugfix/2533-salt-master-kubeconfig
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.5/bugfix/2533-salt-master-kubeconfig
```

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/2533-salt-master-kubeconfig` into
`development/2.4`
* [w/2.5/bugfix/2533-salt-master-kubeconfig](https://github.com/scality/metalk8s/compare/development/2.5...w/2.5/bugfix/2533-salt-master-kubeconfig) will be merged into `development/2.5` 
* [w/2.6/bugfix/2533-salt-master-kubeconfig](https://github.com/scality/metalk8s/compare/development/2.6...w/2.6/bugfix/2533-salt-master-kubeconfig) will be merged into `development/2.6` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @tcarmet

* @thomasdanan

* @sayf-eddine-scality

* @lucieleonard

* @monpote

* @carlito-scality

* @cmonfort

* @nicolast

* @slaperche-scality

* @teddyandrieux

* @nootal

* @gdemonet

* @alexandre-allard-scality

bert-e commented:
# History mismatch



Merge commit #c0f568673263a4319cdf409db7cd38c33aefd21a on the integration branch
`w/2.5/bugfix/2533-salt-master-kubeconfig` is merging a branch which is neither the current
branch `bugfix/2533-salt-master-kubeconfig` nor the development branch
`development/2.5`.

It is likely due to a rebase of the branch `bugfix/2533-salt-master-kubeconfig` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

TeddyAndrieux commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.5/bugfix/2533-salt-master-kubeconfig` with contents from `bugfix/2533-salt-master-kubeconfig`
and `development/2.5`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.5/bugfix/2533-salt-master-kubeconfig origin/development/2.5
 $ git merge origin/bugfix/2533-salt-master-kubeconfig
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.5/bugfix/2533-salt-master-kubeconfig
```

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `bugfix/2533-salt-master-kubeconfig` into
`development/2.4`
* [w/2.5/bugfix/2533-salt-master-kubeconfig](https://github.com/scality/metalk8s/compare/development/2.5...w/2.5/bugfix/2533-salt-master-kubeconfig) will be merged into `development/2.5` 
* [w/2.6/bugfix/2533-salt-master-kubeconfig](https://github.com/scality/metalk8s/compare/development/2.6...w/2.6/bugfix/2533-salt-master-kubeconfig) will be merged into `development/2.6` 



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`




You can set option `create_pull_requests` if you need me to create
**integration pull requests** in addition to integration branches, with:

```
@bert-e create_pull_requests
```

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @jbertran

* @tcarmet

* @thomasdanan

* @sayf-eddine-scality

* @lucieleonard

* @monpote

* @carlito-scality

* @cmonfort

* @nicolast

* @slaperche-scality

* @teddyandrieux

* @nootal

* @gdemonet

* @alexandre-allard-scality

gdemonet commented:
/approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.4`

* :heavy_check_mark: `development/2.5`

* :heavy_check_mark: `development/2.6`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.4`

* :heavy_check_mark: `development/2.5`

* :heavy_check_mark: `development/2.6`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



Please check the status of the associated issue None.

Goodbye gdemonet.
